habitat_baselines/common/obs_transformers.py

```diff
@@ -217,113 +218,496 @@ def from_config(cls, config: Config):
-            mask *= grid[..., 2] == 1
-            grid[
-                ~mask
-            ] = 2  # values bigger than one will be ignored by grid_sample
-            grid_xy = -grid[..., :2].unsqueeze(0)
-            grids.append(grid_xy)
-        grids = torch.cat(grids, dim=0)
+        for input_model in self.input_models:
+            grid, input_mask = input_model.projection(world_pts)
+            # Make sure each point is only assigned to single input
+            input_mask *= not_assigned_mask
+            # Values bigger than one will be ignored by grid_sample
+            grid[~input_mask] = 2
+            # Update not_assigned_mask
+            not_assigned_mask *= ~input_mask
+            grids.append(grid)
+        grids = torch.stack(grids, dim=0)
         return grids
 
-    def _to_equirec(self, batch: torch.Tensor) -> torch.Tensor:
-        """Takes a batch of cubemaps stacked in proper order and converts thems to equirects, reduces batch size by 6"""
+    def _convert(self, batch: torch.Tensor) -> torch.Tensor:
+        """Takes a batch of images stacked in proper order and converts thems, reduces batch size by input_len"""
         batch_size, ch, _H, _W = batch.shape
-        if batch_size == 0 or batch_size % 6 != 0:
-            raise ValueError("Batch size should be 6x")
+        input_len = len(self.input_models)
+        if batch_size == 0 or batch_size % input_len != 0:
+            raise ValueError(f"Batch size should be {input_len}x")
         output = torch.nn.functional.grid_sample(
             batch,
             self._grids_cache,
             align_corners=True,
             padding_mode="zeros",
         )
         output = output.view(
-            batch_size // 6, 6, ch, self.equ_h, self.equ_w
+            batch_size // input_len,
+            input_len,
+            ch,
+            self.output_model.img_h,
+            self.output_model.img_w,
         ).sum(dim=1)
-        return output  # batch_size // 6, ch, self.equ_h, self.equ_w
+        return output  # batch_size // input_len, ch, output_model.img_h, output_model.img_w
 
-    # Convert input cubic tensor to output equirectangular image
-    def to_equirec_tensor(self, batch: torch.Tensor) -> torch.Tensor:
-        batch_size = batch.size()[0]
+    def to_converted_tensor(self, batch: torch.Tensor) -> torch.Tensor:
+        batch_size, ch = batch.size()[:2]
+        input_len = len(self.input_models)
 
-        # Check whether batch size is 6x
-        if batch_size == 0 or batch_size % 6 != 0:
-            raise ValueError("Batch size should be 6x")
+        # Check whether batch size is len(self.input_models) x
+        if batch_size == 0 or batch_size % input_len != 0:
+            raise ValueError(f"Batch size should be {input_len}x")
 
         # to(device) is a NOOP after the first call
         self.grids = self.grids.to(batch.device)
 
+        # Depth conversion
+        if ch == 1 and self.input_zfactors is not None:
```

Comment:
Ah, so one bug here! The semantic sensor could also have a 1 channel output (with uints for object categories), so we need some better way of detecting if it's a depth camera. 
1. We could have it be 1 channel + float. 
2. We could also require by convention that all depth_cameras contain the word "depth" in the UUID. Grr... if we still had the sensor class in scope, would be easy to figure what is a depth sensor. 
3. We could reverse map it, but if a fake_depth ObsTransformer comes first (that generates predicted depth), that wouldn't work either.
4. We could have the user specify it.
5. We could have a separate obs transformer for corrected depth (would inherit with a boolean flag set). Needs to be a distinct name to work around current limitation in config system.
Also we may want to add a logger output under INFO about which channels we are adjusting the depth for.
The ordering of solutions here would be 1 > 4 = 5 > 2 > 3 IMO. Happy to hear your thoughts.

Reply:
More importantly, that if statement may want to be factored out into an "is_depth" function to make it clear for future maintainers what it is suppose to be doing.

Reply:
Ah nvm, saw your post. Yeah, I would say dtype checking is a must since Semantic Segmentation is one of the default sensors in Habitat and was previously supported. Grayscale would be another common use case.

Reply:
IMO, 1 is a hard to find bug waiting to happen.  2 is my preference as requiring that `depth` is in the same of a depth sensor isn't an inconvenient limitation.

Reply:
Give me some time to think about it. Maybe I can try 2 in `ObservationTransformer`, and it passes the `is_depth` flag to `ProjectionConverter` to handle depth correctly.

Reply:
As an extension of 2, we could allow the end user to change the isdepth regex if necessary.